pandas/core/algorithms.py

```diff
@@ -429,7 +429,14 @@ def isin(comps: AnyArrayLike, values: AnyArrayLike) -> np.ndarray:
         # handle categoricals
         return comps.isin(values)  # type: ignore
 
-    comps, dtype = _ensure_data(comps)
+    if hasattr(comps, "dtype") and hasattr(comps.dtype, "name"):
```

Comment:
you need to dive in a bit here to see exactly where something is incorrect. We need maintainable general code, not specific references as far as possible.

Reply:
i was searching today for some resolution and i sure that the problems is that you cannot do the dtype=dtype, cause this converts an array with dtype 'object' to an array with the same type as the one that entered. And i know that if we convert both datas to dtype=object it works perfect with dataframes, but i think that breaks when using isin with arrays. My question is: Should be fine use isin with arrays, or just with dataframes?

Reply:
I think i got the problem! The real problem is at the function that compares the values: f = htable.ismember_object, this expect to receive always the same type like always int64, float64 or object, i dont know if the way that is implemented is right

Reply:
The problem is not at the _ensure_data func, but in the fact that for comparing the dataframe and the isin both need to be same type. I think that the way i have implemented solve a lot of the problems with that. I dont think there is a soluting better than that at this point, but if you think its no good im going to keep trying to find a way

Reply:
as i said need to dive in here a little more; this is a very heavily used function and we need to have the code in a good location w/o this type of very specific type checking (which is already done but elsewhere)

Reply:
Alright, i was trying to look where is the problem and i think that a solution is to make a algorithm that works when comps has dtype = int and values dtype = object. Because today we have 3 algorithms: htable.ismember_object(both has type Object), htable.ismember_int64 (both are int64) and htable.ismember_float64 (both are float64). So when you have a dtype Object this is casted to the df dtype and send to the algorithm. But I think i cant do this changes by myself. I think that i can make some tests like the one i made to identify that this error and at least the people will know that this error exists, what do you think?

Reply:
so you could add tests that are xfailed

Reply:
thats what i though, can i send them here on this pull request?

Reply:
sure (but leave the issue open, e.g. changes closes -> xref in the PR top)

Reply:
i have just added some xfailed tests with differents dtype, I identify another bug that np.nan is the same as "NaN" in isin func, so i added this too. Can you check if everything is right?